### ISSUE TITLE
Combining attributes from usage and the template

### ISSUE THREAD

**Issue by MrHayato** [aurelia/templating#344]
We have a custom attribute that we've made for local css modules (using webpack's css-loader's `modules` option) which maps the class in the css definition to the local identifier that's generated. The custom attribute is used like so:

```html
<header css-class="header">
  <h1 css-class="heading ${isRed ? 'red' : 'black'}">My Heading</h1>
</header>
```

This works great; it takes the values in css-class and gets the mapped identifier, and then we apply the class to it accordingly.

The problem is, we also want to be able to apply the attribute to the `<template>` tag for the component's root level element. This works, but if we also apply a css-class attribute when the component is used, it is overridden instead of combined. For classes, the values are combined, and we'd like the same for our custom attribute. Is there a setting that I'm missing? I can't seem to find it when going through the source.

Thanks!

**Comment by jdanyow** [comment]
The binding system uses the [`ClassObserver`](https://github.com/aurelia/binding/blob/master/src/class-observer.js) when making changes to an html element's `class` attribute.  The `ClassObserver` has logic that ensures a particular binding's class-list changes don't interfere with another binding's class-list changes.

Try changing your custom attribute to something like this:

```js
import {inject, ObserverLocator} from 'aurelia-framework';

@inject(Element, ObserverLocator)
export class CssClassCustomAttribute {
  constructor(element, observerLocator) {
    this.observer = observerLocator.getObserver(element, 'class');  // this will retrieve an instance of ClassObserver
  }

  valueChanged(newValue, oldValue) {
    // ... css modules logic ... eg "foo bar baz" becomes "foo-123 bar-456 baz-789" ...

    // use the ClassObserver to apply the changes.
    this.observer.setValue(newValue);
  }
}
```

**Comment by MrHayato** [comment]
Hmmm, I'm not sure that's what I want. Setting the class itself isn't an issue, it's the css-class attribute getting overridden. Here's a better example:

```html
<!-- my-comp.html -->
<template css-class="local-class1">
    <div>Stuff</div>
</template>
```

```html
<!-- index.html -->
<template>
  <require from="./my-comp.html"></require>
  <my-comp css-class="my-other-class"></my-comp>
</template>
```

In the scenario above, I would like the resulting `<my-comp>` once rendered, to look like: `<my-comp css-class="my-other-class local-class1"></my-comp>`.

Would this require me to create a custom observer?

**Comment by jdanyow** [comment]
I don't think that will be the answer in this case- will need to defer to @EisenbergEffect on this one.

**Comment by MrHayato** [comment]
@EisenbergEffect Any input on this? Running into issues with some of our components and not sure how to proceed.

**Comment by EisenbergEffect** [comment]
I'm not sure we can handle this directly because css and styles are really a special case. We know how to combine those. However, we don't know anything about css-class attribute, how it works. So, we assume that the consumer attr overrides the one inside the template. I'm not sure what a good way to address this would be. We can't assume they should be merged.

**Comment by MrHayato** [comment]
I definitely agree that it shouldn't be assumed to be merged. Our usage of it just happens to be very similar to the way class attributes work. Is there something that can be overridden or a plugin that I can create to change the behavior such that they get merged?

**Comment by EisenbergEffect** [comment]
@MrHayato Take a look at this line of code: https://github.com/aurelia/templating/blob/master/src/view-factory.js#L236

You will notice that there's literally a missing `else` case there because we don't know what to do if we've found values in both cases. If you would like to propose a solution to that, we'd love to work with you on that.

**Comment by MrHayato** [comment]
What about providing the behaviour from the @customAttribute decorator?

```
@customAttribute('my-attribute', { surrogateBehavior: 'merge' })
export class MyAttribute {}
```

Or alternatively, instead of a `'merge'` string, it could be a custom SurrogateBehaviour class that has an `apply(element, attributes)`?

**Comment by EisenbergEffect** [comment]
I don't think we could just generically "merge" I think we would need you to provide a merge function so that you could do the work and return the final value back since "merging" could mean different things depending on the attribute.

**Comment by MrHayato** [comment]
Sorry, I typed things out in a hurry because I was swamped with stuff at work. This is more what I had in mind when I typed the `merge` value:

```
@customAttribute('my-attribute', { surrogateBehavior: 'merge' })
export class MyAttribute {
    merge(element, oldValue, newValue) {
        // Specify custom merge functionality here. In this custom attribute's case, we'll combine them.
        element.setAttribute('my-attribute', oldValue + ' ' + newValue);
    }
}
```
So the `merge` string is just the function that would be called to do the merge.

**Comment by EisenbergEffect** [comment]
Hmm. I'll need to think about that. It's tricky since the attribute classes haven't been instantiated at that point.

**Comment by Garrett-Smith-iq** [comment]
@EisenbergEffect Any chance you've given this issue any more thought?

**Comment by EisenbergEffect** [comment]
It hasn't been a high priority. If someone wants to work on a PR to fix this, we would certainly help to get that in.

**Comment by Alexander-Taran** [comment]
stale

**Comment by Alexander-Taran** [comment]
```
async triageQuestion(id) =>{
 respondSameDay()
 let response = await getResponseFromOP( { timeout : 4weeks } )
 if(!response) {
    closeQuestion(id,randomReason())
 }
}
```

**Comment by EisenbergEffect** [comment]
I think this is something that @fkleuver should review WRT to vNext surrogate behavior.

**Comment by fkleuver** [comment]
@EisenbergEffect what was the issue with @WillsonHaw 's suggestion?
```ts
@customAttribute('my-attribute', { surrogateBehavior: 'merge' })
export class MyAttribute {
    merge(element, oldValue, newValue) {
        // Specify custom merge functionality here. In this custom attribute's case, we'll combine them.
        element.setAttribute('my-attribute', oldValue + ' ' + newValue);
    }
}
```

What do you mean by `It's tricky since the attribute classes haven't been instantiated at that point.`?

**Comment by EisenbergEffect** [comment]
If I recall, the code for this was in the compiler, so the behaviors wouldn't be instanced at that time. It would probably have to be a static method instead to make it work in that case. You're the expert on what the best way to handle this would be in vNext.
